src/xmipp/libraries/reconstruction/subtract_projection.cpp

```diff
@@ -0,0 +1,362 @@
+ #include <vector>
+ #include <utility>
+
+ // Read arguments ==========================================================
+ void ProgSubtractProjection::readParams()
+ {
+	fnParticles = getParam("-i");
+ 	fnVolR = getParam("--ref");
+	fnOut=getParam("-o");
+	if (fnOut=="")
+		fnOut="output_particle_";
+	fnMask=getParam("--mask");
+	fnMaskVol=getParam("--maskVol");
+	iter=getIntParam("--iter");
+	sigma=getIntParam("--sigma");
+	cutFreq=getDoubleParam("--cutFreq");
+	lambda=getDoubleParam("--lambda");
+	fnProj=getParam("--saveProj");
+ }
+
+ // Show ====================================================================
+ void ProgSubtractProjection::show() const{
+    if (!verbose)
+        return;
+	std::cout<< "Input particles:\t" << fnParticles << std::endl
+	<< "Reference volume:\t" << fnVolR      << std::endl
+	<< "Volume mask:\t" << fnMaskVol   << std::endl
+	<< "Mask:\t" << fnMask      << std::endl
+	<< "Sigma:\t" << sigma       << std::endl
+	<< "Iterations:\t" << iter        << std::endl
+	<< "Cutoff frequency:\t" << cutFreq     << std::endl
+	<< "Relaxation factor:\t" << lambda      << std::endl
+	<< "Output particles:\t" << fnOut 	     << std::endl;
+ }
+
+ // usage ===================================================================
+ void ProgSubtractProjection::defineParams()
+ {
+     //Usage
+     addUsageLine("");
+     //Parameters
+     addParamsLine("-i <particles>\t: Particles metadata (.xmd file)");
+     addParamsLine("--ref <volume>\t: Reference volume to subtract");
+     addParamsLine("[-o <structure=\"\">]\t: Output filename suffix for subtracted particles");
+     addParamsLine("\t: If no name is given, then output_particles.xmd");
+     addParamsLine("[--maskVol <maskVol=\"\">]\t: 3D mask for input volume");
+     addParamsLine("[--mask <mask=\"\">]\t: 3D mask for the region of subtraction");
+     addParamsLine("[--sigma <s=3>]\t: Decay of the filter (sigma) to smooth the mask transition");
+     addParamsLine("[--iter <n=1>]\t: Number of iterations");
+     addParamsLine("[--cutFreq <f=0>]\t: Cutoff frequency (<0.5)");
+     addParamsLine("[--lambda <l=0>]\t: Relaxation factor for Fourier Amplitude POCS (between 0 and 1)");
+	 addParamsLine("[--saveProj <structure=\"\"> ]\t: Save subtraction intermediate files (projection adjusted)");
+     addExampleLine("A typical use is:",false);
+     addExampleLine("xmipp_subtract_projection -i input_particles.xmd --ref input_map.mrc --maskVol mask_vol.vol --mask mask.vol -o output_particles --iter 5 --lambda 1 --cutFreq 0.44 --sigma 3");
+ }
+
+ void POCSmaskProj(const MultidimArray<double> &mask, MultidimArray<double> &I) {
+ 	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(I)
+ 	DIRECT_MULTIDIM_ELEM(I,n)*=DIRECT_MULTIDIM_ELEM(mask,n);
+ }
+
+ void POCSFourierAmplitudeProj(const MultidimArray<double> &A, MultidimArray< std::complex<double> > &FI, double lambda, const MultidimArray<double> &rQ, int Isize) {
+ 	int Isize2 = Isize/2;
+ 	double Isizei = 1.0/Isize;
+ 	double wx;
```

Comment:
preferably declare variable immediately before you need it, in this case before line 110
(multiple occurrences)

Reply:
If I do what you propose here, the variable will be declared once for each iteration, will it be correct??

Reply:
yes, don't worry about those. Don't do it for 'big objects', like images and such.